core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -0,0 +1,1077 @@
+   * count permitting, it will partition the input set of {@link Sequence} to do 2 layer parallel merge.
+   *
+   * For the first layer, the partitions of input sequences are each wrapped in {@link YielderBatchedResultsCursor}, and
+   * for each partition a {@link PrepareMergeCombineInputsAction} will be executed to to wait for each of the yielders to
+   * yield {@link ResultBatch}. After the cursors all have an initial set of results, the
+   * {@link PrepareMergeCombineInputsAction} will execute a {@link MergeCombineAction}
+   * to perform the actual work of merging sequences and combining results. The merged and combined output of each
+   * partition will itself be put into {@link ResultBatch} and pushed to a {@link BlockingQueue} with a
+   * {@link ForkJoinPool} {@link QueuePusher}.
+   *
+   * The second layer will execute a single {@link PrepareMergeCombineInputsAction} to wait for the {@link ResultBatch}
+   * from each partition to be available in their 'output' {@link BlockingQueue} which each is wrapped in
+   * {@link BlockingQueueuBatchedResultsCursor}. Like the first layer, after the {@link PrepareMergeCombineInputsAction}
+   * is complete and some {@link ResultBatch} are ready to merge from each partition, it will execute a
+   * {@link MergeCombineAction} do a final merge combine of all the parallel computed results, again pushing
+   * {@link ResultBatch} into a {@link BlockingQueue} with a {@link QueuePusher}.
+   */
+  private static class MergeCombinePartitioningAction<T> extends RecursiveAction
+  {
+    private final List<Sequence<T>> sequences;
+    private final Ordering<T> orderingFn;
+    private final BinaryOperator<T> combineFn;
+    private final BlockingQueue<ResultBatch<T>> out;
+    private final int queueSize;
+    private final int parallelism;
+    private final int yieldAfter;
+    private final int batchSize;
+    private final boolean hasTimeout;
+    private final long timeoutAt;
+    private final CancellationGizmo cancellationGizmo;
+
+    private MergeCombinePartitioningAction(
+        List<Sequence<T>> sequences,
+        Ordering<T> orderingFn,
+        BinaryOperator<T> combineFn,
+        BlockingQueue<ResultBatch<T>> out,
+        int queueSize,
+        int parallelism,
+        int yieldAfter,
+        int batchSize,
+        boolean hasTimeout,
+        long timeoutAt,
+        CancellationGizmo cancellationGizmo
+    )
+    {
+      this.sequences = sequences;
+      this.combineFn = combineFn;
+      this.orderingFn = orderingFn;
+      this.out = out;
+      this.queueSize = queueSize;
+      this.parallelism = parallelism;
+      this.yieldAfter = yieldAfter;
+      this.batchSize = batchSize;
+      this.hasTimeout = hasTimeout;
+      this.timeoutAt = timeoutAt;
+      this.cancellationGizmo = cancellationGizmo;
+    }
+
+    @Override
+    protected void compute()
+    {
+      try {
+        final int parallelTaskCount = computeNumTasks();
```

Comment:
what if multiple queries arrive concurrently, there would be races and they might overestimate available parallel capacity ?

Reply:
currently it can definitely overestimate how parallel it can go, though in my stress testing so far it doesn't seem to actually matter that much, as each fork join pool task runs for a short period so that the work is still all interleaved, just done slightly slower since there are more tasks to run.

If we wanted to try and only allocate as much parallelism as there are threads in the pool then we would need an external mechanism to share between all queries I think beyond asking the fork join pool directly, that grabs a number of slots from a fixed size counter and returns them when completed. Maybe something like the thread holders in @jihoonson previous attempt at parallel merges. This is because asking the fork join pool itself isn't adequate, due to the nature of how this operates where there is only 1 task per query at the time this is computed, but after that initial task is done it then executes all of the parallel tasks to do the actual merge work. Like imagine 3 queries come in, it launches 3 tasks to determine how parallel to go, and all ask the pool what the usage is and only see the 3 tasks, so they each think they can use all the remaining capacity of the pool.

Anyway, I was hoping to defer doing something like this until it becomes obvious that we actually need it, which I'm not certain that we do yet.

Reply:
All that said, i certainly there are smarter ways we could compute the level of parallelism and also how the input sequences are partitioned, the current approach is very greedy.

Reply:
thanks , glad to know that it has been given considerable thought.

can we emit a metric for number of sequences handled per partition i.e. `sequences.size()/partitions` , as this code runs, It will be nice to be able to have visibility into this.

Reply:
:100: on board with adding metrics, I really want them to help me tune it, have just been on the fence if I should do it in this PR or a follow-up.

I'm still thinking about what all I would like to see, definitely partition sizes, number of partitions, number of total tasks required to fully process the result sets (recursion depths), maybe more, though I need to weigh this probably with what is useful to help me tune stuff now with what is actually useful long term, which would probably be translating these things I want to see into something an operator could use to tell how saturated their brokers are to aid in scaling decisions.

Reply:
yes, my proposed metric was to see, for me as user, basically how much parallelism is really happening. actually I would add one more, periodically, number of queries received and number of queries parallelized.

I am OK if you wanna do that in a later PR , it should just be done before this code is shipped in a release.
more metrics could of course be there for devs to learn about reality and change code based on that.

on the scaling (I'm assuming you mean increasing number of broker boxes), I think the decision should be made based on much higher level metrics such as cpu/memory utilization of the boxes etc , but anyways that is speculation at this point . that said, I did write a k8s operator which I am hoping to open source soon (currently it only does the deployment) :)